pkg/controller/lease.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// podmaster is a simple utility, it attempts to acquire and maintain a lease-lock from etcd using compare-and-swap.
+// if it is the master, it copies a source file into a destination file.  If it is not the master, it makes sure it is removed.
+//
+// typical usage is to copy a Pod manifest from a staging directory into the kubelet's directory, for example:
+//   podmaster --etcd-servers=http://127.0.0.1:4001 --key=scheduler --source-file=/kubernetes/kube-scheduler.manifest --dest-file=/manifests/kube-scheduler.manifest
+package controller
+
+import (
+	"strings"
+	"time"
+
+	"github.com/coreos/go-etcd/etcd"
+	"github.com/golang/glog"
+	etcdstorage "k8s.io/kubernetes/pkg/storage/etcd"
+)
+
+type Config struct {
+	EtcdServers string
+	Key         string
+	whoami      string
+	ttl         uint64
+	sleep       time.Duration
+	lastLease   time.Time
+	Running     func() bool
+	Lease       func() bool
+	Unlease     func() bool
+}
+
+// runs the election loop. never returns.
+func (c *Config) leaseAndUpdateLoop(etcdClient *etcd.Client) {
+	for {
+		master, err := c.acquireOrRenewLease(etcdClient)
+		if err != nil {
+			glog.Errorf("Error in master election: %v", err)
+			if uint64(time.Now().Sub(c.lastLease).Seconds()) < c.ttl {
+				continue
+			}
+			// Our lease has expired due to our own accounting, pro-actively give it
+			// up, even if we couldn't contact etcd.
+			glog.Infof("Too much time has elapsed, giving up lease.")
+			master = false
+		}
+		if err := c.update(master); err != nil {
+			glog.Errorf("Error updating files: %v", err)
+		}
+		time.Sleep(c.sleep)
+	}
+}
+
+// acquireOrRenewLease either races to acquire a new master lease, or update the existing master's lease
+// returns true if we have the lease, and an error if one occurs.
+// TODO: use the master election utility once it is merged in.
+func (c *Config) acquireOrRenewLease(etcdClient *etcd.Client) (bool, error) {
```

Comment:
My basic testing with watch through kubectl was troubling.  I got updates for changes, but no update when the lock was deleted.  Watch based would be best imo, but it may not work.